src/backend/distributed/connection/shared_connection_stats.c

```diff
@@ -800,14 +919,54 @@ ShouldWaitForConnection(int currentConnectionCount)
 }
 
 
+static SharedWorkerNodeConnStatsHashKey
+PrepareWorkerNodeHashKey(const char *hostname, int port)
+{
+	SharedWorkerNodeConnStatsHashKey key;
+	strlcpy(key.hostname, hostname, MAX_NODE_LENGTH);
+	if (strlen(hostname) > MAX_NODE_LENGTH)
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("hostname exceeds the maximum length of %d",
+							   MAX_NODE_LENGTH)));
+	}
+	key.port = port;
+	return key;
+}
+
+
+static SharedWorkerNodeDatabaseConnStatsHashKey
+PrepareWorkerNodeDatabaseHashKey(const char *hostname,
+								 int port,
+								 Oid database)
+{
+	SharedWorkerNodeDatabaseConnStatsHashKey workerNodeDatabaseKey;
+	workerNodeDatabaseKey.workerNodeKey = PrepareWorkerNodeHashKey(hostname, port);
+	workerNodeDatabaseKey.database = database;
+	return workerNodeDatabaseKey;
+}
+
+
 static uint32
 SharedConnectionHashHash(const void *key, Size keysize)
 {
-	SharedConnStatsHashKey *entry = (SharedConnStatsHashKey *) key;
+	SharedWorkerNodeConnStatsHashKey *entry = (SharedWorkerNodeConnStatsHashKey *) key;
 
 	uint32 hash = string_hash(entry->hostname, NAMEDATALEN);
 	hash = hash_combine(hash, hash_uint32(entry->port));
-	hash = hash_combine(hash, hash_uint32(entry->databaseOid));
+
+	return hash;
+}
+
+
+static uint32
+SharedWorkerNodeDatabaseHashHash(const void *key, Size keysize)
+{
+	SharedWorkerNodeDatabaseConnStatsHashKey *entry =
+		(SharedWorkerNodeDatabaseConnStatsHashKey *) key;
+	uint32 hash = string_hash(entry->workerNodeKey.hostname, NAMEDATALEN);
+	hash = hash_combine(hash, hash_uint32(entry->workerNodeKey.port));
```

Comment:
nit: Would be good to call SharedConnectionHashHash here instead, instead of re-implementing it. Same for the Compare function.